### PR TITLE
Annotate completion providers

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

File: docs/Breaking API Changes.md [hunk 1]
```diff
@@ -46,3 +46,10 @@ The change affects compatibility in two ways:
 - DateTimeConstant(-1) will still count when we check that you don’t specify two default values. The compiler will produce an error, instead of succeeding (and producing IL with two attributes).
 
 PR: https://github.com/dotnet/roslyn/pull/11536
+
+# Version 4.1.0
```
Comment (Youssef1313) [inline docs/Breaking API Changes.md:50]: This document is mostly outdated. https://github.com/dotnet/roslyn/issues/54242 is tracking documenting API breaking changes. But no conclusion yet it seems.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs [hunk 2]
```diff
@@ -90,7 +90,7 @@ internal SyntaxTriviaList ResidualTrivia
                     }
                     else
                     {
-                        return default(SyntaxTriviaList);
+                        return default;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs:93]: These changes in the SyntaxNodeRemover seem unrelated to annotation. Consider reverting so you don't need compiler reviews on this PR.

File: src/Tools/ExternalAccess/OmniSharp/Completion/OmniSharpCompletionService.cs [hunk 3]
```diff
@@ -23,7 +23,7 @@ internal static class OmniSharpCompletionService
             CancellationToken cancellationToken = default)
             => completionService.GetCompletionsInternalAsync(document, caretPosition, trigger, roles, options, cancellationToken);
 
-        public static string GetProviderName(this CompletionItem completionItem) => completionItem.ProviderName;
+        public static string? GetProviderName(this CompletionItem completionItem) => completionItem.ProviderName;
```
Comment (333fred) [inline src/Tools/ExternalAccess/OmniSharp/Completion/OmniSharpCompletionService.cs:26]: This change looks good to me. Luckily, while it definitely surprises me, I wasn't depending on provider name being non-null in O#'s service.

**Review by 333fred — APPROVED** [review]
Omnisharp change LGTM

File: src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs (hunk 2 above)
Comment (tmat) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs:93]: Accident. Reverted.

File: src/Features/Core/Portable/Completion/Providers/AbstractDocCommentCompletionProvider.cs [hunk 4]
```diff
@@ -270,19 +268,8 @@ private static string FormatParameterRefTag(string kind, string name)
...
-                includesCommitCharacter = false;
-            }
-            else
-            {
-                beforeCaretText = XmlDocCommentCompletionItem.GetBeforeCaretText(item);
-                afterCaretText = XmlDocCommentCompletionItem.GetAfterCaretText(item);
-            }
-
+            var beforeCaretText = XmlDocCommentCompletionItem.GetBeforeCaretText(item);
+            var afterCaretText = XmlDocCommentCompletionItem.GetAfterCaretText(item);
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Providers/AbstractDocCommentCompletionProvider.cs:272]: The behavior triggered on space that was relying on this code was removed in [https://github.com/dotnet/roslyn/pull/21479](https://github.com/dotnet/roslyn/pull/21479). Just want to point it out in case someone else who's reviewing this PR also wondered about this

Comment (tmat) [reply src/Features/Core/Portable/Completion/Providers/AbstractDocCommentCompletionProvider.cs:272]: Good to know. Thanks for the ptr.

File: src/Features/Core/Portable/Completion/Providers/AbstractOverrideCompletionProvider.ItemGetter.cs [hunk 5]
```diff
@@ -58,7 +57,10 @@ private partial class ItemGetter
                 CancellationToken cancellationToken)
             {
                 var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
                 var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Providers/AbstractOverrideCompletionProvider.ItemGetter.cs:61]: `GetRequiredSyntaxTreeAsync`?

File: src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs [hunk 6]
```diff
@@ -208,10 +206,10 @@ private static Document FindAppropriateDocumentForDescriptionContext(Document do
                 }
             }
 
-            return contextDocument;
+            return contextDocument ?? document;
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/Providers/SymbolCompletionItem.cs:209]: I wonder if this would cover up some bug in the future, since we do expect `GetDocument(contextId)` above to return non-null. Maybe immediately return the value from a call to `GetRequiredDocument(contextId)` in the if-statement instead?

**Review by genlu — APPROVED** [review]
:shipit:
